crates/matrix-sdk/src/sliding_sync/client.rs

```diff
@@ -178,16 +274,137 @@ async fn update_in_memory_caches(client: &Client, response: &SyncResponse) -> Re
 mod tests {
     use std::collections::BTreeMap;
 
+    use assert_matches::assert_matches;
     use matrix_sdk_base::notification_settings::RoomNotificationMode;
     use matrix_sdk_test::async_test;
     use ruma::{assign, room_id, serde::Raw};
     use serde_json::json;
+    use url::Url;
 
     use crate::{
         error::Result, sliding_sync::http, test_utils::logged_in_client_with_server,
         SlidingSyncList, SlidingSyncMode,
     };
 
+    use super::{discover_homeserver, get_supported_versions, Error, Version, VersionBuilder};
+
+    #[test]
+    fn test_version_builder_none() {
+        assert_matches!(VersionBuilder::None.build(None, None), Ok(Version::None));
+    }
+
+    #[test]
+    fn test_version_builder_proxy() {
+        let expected_url = Url::parse("https://matrix.org:1234").unwrap();
+
+        assert_matches!(
+            VersionBuilder::Proxy { url: expected_url.clone() }.build(None, None),
+            Ok(Version::Proxy { url }) => {
+                assert_eq!(url, expected_url);
+            }
+        );
+    }
+
+    #[test]
+    fn test_version_builder_native() {
+        assert_matches!(VersionBuilder::Native.build(None, None), Ok(Version::Native));
+    }
+
+    #[test]
+    fn test_version_builder_discover_proxy() {
+        let expected_url = Url::parse("https://matrix.org:1234").unwrap();
+        let mut response = discover_homeserver::Response::new(
+            discover_homeserver::HomeserverInfo::new("matrix.org".to_owned()),
+        );
+        response.sliding_sync_proxy =
+            Some(discover_homeserver::SlidingSyncProxyInfo::new(expected_url.to_string()));
+
+        assert_matches!(
+            VersionBuilder::DiscoverProxy.build(Some(&response), None),
+            Ok(Version::Proxy { url }) => {
+                assert_eq!(url, expected_url);
+            }
+        );
+    }
+
+    #[test]
+    fn test_version_builder_discover_proxy_no_well_known() {
+        assert_matches!(
+            VersionBuilder::DiscoverProxy.build(None, None),
+            Err(Error::VersionCannotBeDiscovered(msg)) => {
+                assert_eq!(msg, "`.well-known` is `None`");
```

Comment:
Maybe worth creating a small error type with thiserror, to avoid stringly-typed errors?